compiler/rustc_mir/src/transform/instcombine.rs

```diff
@@ -111,6 +140,85 @@ impl OptimizationFinder<'b, 'tcx> {
         }
     }
 
+    fn find_eq_not(&mut self, rvalue: &Rvalue<'tcx>, location: Location) -> Option<()> {
+        // Optimize the sequence
+        // _4 = Eq(move _5, const 2_u8);
+        // StorageDead(_5);
+        // _3 = Not(move _4);
+        //
+        // into _3 = Ne(move _5, const 2_u8)
+        if let Rvalue::UnaryOp(UnOp::Not, op) = rvalue {
+            let place = op.place()?;
+            // See if we can find a Eq that assigns `place`.
+            // We limit the search to 3 statements lookback.
+            // Usually the first 2 statements are `StorageDead`s for operands for Eq.
+            // We record what is marked dead so that we can reorder StorageDead so it comes after Ne
+
+            // We will maximum see 2 StorageDeads
+            let mut seen_storage_deads: SmallVec<[_; 2]> = SmallVec::new();
+            let lower_index = location.statement_index.saturating_sub(3);
+            for (stmt_idx, stmt) in self.body.basic_blocks()[location.block].statements
+                [lower_index..location.statement_index]
+                .iter()
+                .enumerate()
+                .rev()
+            {
+                match &stmt.kind {
+                    rustc_middle::mir::StatementKind::Assign(box (l, r)) => {
+                        if *l == place {
+                            match r {
+                                Rvalue::BinaryOp(BinOp::Eq, op1, op2) => {
+                                    // We need to make sure that the StorageDeads we saw are for
```

Comment:
I don't understand why this is important. I'd rather think that the `StorageDead` for `l` is important, otherwise we should not remove the `Neg`, but instead also copy the result of the `Ne` operation to `l`.

Reply:
Can you expand on your statement? I don't understand your point about not removing `Neq` (you mean Not? or Ne?).

Now that I think about it, yeah, it is perhaps unneccessary to require that the StorageDeads are for the Eq operands. Hmm do we even need to do anything special about potential StorageDeads between Eq and Not? I think not, besides doing the swap, which I think we can just do unconditionally for simplicity. We already assert that only StorageDead statements can appear between Eq and Not, and I think simply swapping a StorageDead for a complete unrelated local with Ne should not matter.

Reply:
What I'm wondering is whether it is generally simpler to replace

```rust
                                    // _4 = Eq(move _5, move _6);
                                    // StorageDead(_5);
                                    // StorageDead(_6);
                                    // _3 = Not(move _4);
```

with

```rust
                                    // _4 = Ne(move _5, move _6);
                                    // StorageDead(_5);
                                    // StorageDead(_6);
                                    // _3 = move _4;
```

since then you don't have to care about any liveness or similar and leave further changes to other optimizations.

Reply:
If you want to get rid of the additional `_3 = move 4;`, maybe another peephole optimization may be better? One that just checks for `_x = foo; /*arbitrary statements not touching _x or _y*/ _y = _x;` and changes it to `_y = foo; /*arbitrary statements not touching _x or _y*/ nop;`

Reply:
I think it may be beneficial to keep the current version. With your proposal, the interaction in https://github.com/rust-lang/rust/pull/77031#discussion_r492364145 would break. This is only one case, yeah, but it still seems worth to keep that.

What do you think?

Reply:
I personally would prefer to break that interaction and fix it in a separate PR. Otherwise we need to implement both the simpler version of this PR together with the peephole opt for trivial copy propagation.